src/Microsoft.AspNetCore.Sockets.Common/Internal/ChannelConnection.cs

```diff
@@ -7,21 +7,33 @@ namespace Microsoft.AspNetCore.Sockets.Internal
 {
     public static class ChannelConnection
     {
+        public static ChannelConnection<T, U> Create<T, U>(Channel<T> input, Channel<U> output)
+        {
+            return new ChannelConnection<T, U>(input, output);
+        }
+
         public static ChannelConnection<T> Create<T>(Channel<T> input, Channel<T> output)
```

Comment:
You can remove this. If the type is the same, the overload above will work and inference will just fill in both parameters with the same type. That would also mostly eliminate the need for `ChannelConnection<T>`.